### DISCUSSION TITLE
Autogen And Semantic Kernels Using python

### DISCUSSION THREAD

**Discussion by lovedeepatsgit** [microsoft/semantic-kernel#9983] [General]
Hi Everyone, I have worked with both Autogen and semantic kernels but I have worked with these two to only work with muti-agent approaches. Firstly, I have started with Autogen and it was a great experience, then Semantic Kernels has also launched their Agentic Framework, And on Agents session on Ignite I have came to know that Autogen is for Ideation and Semantic Kernels is for production https://techcommunity.microsoft.com/blog/azure-ai-services-blog/introducing-azure-ai-agent-service/4298357. But I with my experience I got that Semantic Kernels are more complex as compared to Autogen, because In Autogen We can directly create Agents but in SK first there is a need to create kernel, then Ai Service, then various settings like Prompt Execution Settings, and then We can create a Agent. And Apart from that Autogen offers sequential Chat, multi chat directly, but in SK further Selection and Termination functions are needed to be created. So, what are the reasons due to which I should use SK over Autogen. There is saying about production and commercials functions in SK, So what actually are these function that makes SK superior to Autogen? Or What are the cases in which SK should be the first choice upon Autogen.

**Comment by moonbox3** [comment]
Hi @lovedeepatsgit, thanks for your question. 

AutoGen is an awesome framework -- I agree! Its goal is to push the boundaries of the latest and greatest agentic research and patterns. AutoGen is developed by a team at Microsoft Research, and while it’s not at General Availability (GA), this allows them to iterate quickly. In contrast, Semantic Kernel is at GA (v1.0+ for all three kernels—Python, C#, and Java). Our goal with Semantic Kernel is to provide a robust, stable, and enterprise-ready SDK that enables app developers to leverage the latest AI services, models, connectors, memory, agents, processes, and more in a safe and reliable way. 

We work closely with the AutoGen team to integrate their most successful agentic research patterns into Semantic Kernel. These features go through a "graduation" process: starting as experimental, then moving to preview, and finally reaching general availability (GA). (Note: our SK agent framework should be at GA in early 2025.)

You're absolutely right that we're helping position AutoGen as a tool for proof of concepts, experimentation, and research. At the same time, we’re bridging the gap between the two frameworks, enabling AutoGen users to transition their agents into Semantic Kernel when they’re ready for production scenarios. See this blog post [Microsoft’s Agentic AI Frameworks: AutoGen and Semantic Kernel](https://devblogs.microsoft.com/semantic-kernel/microsofts-agentic-ai-frameworks-autogen-and-semantic-kernel/).

---

**Your Comment on Complexity**

Regarding your comment: 

> "_With my experience, I found Semantic Kernel more complex compared to AutoGen. In AutoGen, we can directly create agents, but in SK, we need to first create a kernel, then an AI service, various settings like Prompt Execution Settings, and only then can we create an agent._"

This is great feedback -- thank you! Let me elaborate:

- Semantic Kernel provides two agent "flavors": **Chat Completion Agent** and **Assistant Agent**.  
  - **Chat Completion Agent** closely follows the Chat Completion construct. You create an AI service, and optionally a kernel and/or plugins registered on that kernel. From there, you can create the agent and manage chat history as part of the agent's interactions.  
    - At a minimum you do need an AI service to drive API calls to the model, but you don't need to include a Kernel or plugins.
  - **Assistant Agent** uses the OpenAI Assistant v2 APIs. This approach allows for server-side assistant management with chat handled via threads. It also supports plugins/functions and provides additional capabilities like the code interpreter and OpenAI's vector store.
    - For this type of agent, you do not need to specify an AI service because you create the assistant either for Azure or OpenAI. Additionally, you do not need to include a kernel/plugins if you don't need that behavior. You can purely stick to the code interpreter / file store vector search with a few keyword arguments.
  - Both agents allow developers to tailor the agent to specific needs, and even combine different agent types in group chats.  

---

**Your Question: Why Use SK Over AutoGen?**

> "_What are the reasons to use SK over AutoGen? What production or commercial functions make SK superior? In what scenarios should SK be the first choice over AutoGen?_"

Here’s why Semantic Kernel might be the better choice for specific use cases:

- **Version Stability**: All SK kernels are at v1.0+, meaning no breaking changes are introduced in major or minor version updates. This ensures enterprise developers can build robust applications with confidence, knowing their dependencies will remain stable. In contrast, AutoGen’s research focus and pre-GA status mean it may introduce breaking changes in the future.  

- **Enterprise Readiness**: Semantic Kernel is designed for production scenarios. It’s already used by many customers in production environments. You can check out case studies and success stories on our [developer blog](https://devblogs.microsoft.com/semantic-kernel/).  

- **Feature Maturity**: Semantic Kernel’s features, once integrated, go through the aforementioned graduation process, ensuring they’re production-ready. This makes it ideal for applications requiring reliability, security, and scalability.  

- **Process Framework for Advanced Workflow Design**: Semantic Kernel includes a **Process Framework** that enables developers to create robust, stateful workflows with repeatable, graph-based interactions.  
  - The framework allows you to define processes as connected steps where each step executes Semantic Kernel functions, while edges manage dependencies and events between steps. This is particularly useful for building complex applications, like multi-turn conversations or intricate decision trees, that require stateful management.  
  - With Python support recently added, it’s now easier than ever to create and manage processes using the [Process Framework samples](https://github.com/microsoft/semantic-kernel/tree/main/python/samples/getting_started_with_processes).  
  - Learn more about how the Process Framework works on our [developer blog](https://devblogs.microsoft.com/semantic-kernel/process-framework-gets-python-support-and-more/).  

AutoGen excels in experimentation, prototyping, and exploring cutting-edge research. Semantic Kernel, however, is purpose-built for enterprise-grade applications that require a stable, long-term solution.

---

I hope this provides clarity and addresses your concerns. Feel free to reach out with further questions or feedback.

**Reply by lovedeepatsgit** [reply]
Thanks @moonbox3, But I have one more question. Autogen is completely based on python, and Semantic Kernels is available in C#, Python and Java, But in SK more functions are available using C# (.NET), as compared to python. What is the reason behind this, because if SK is taking functions from Autogen Which is in python, Then in SK more and rapidly functions should come in Python Version

**Reply by moonbox3** [reply]
AutoGen has .Net capability, too (https://github.com/microsoft/autogen/tree/main/dotnet). 

We strive for parity between SK languages, but there are times where there can be some divergences (for example, if one language goes first with a feature and the other language has resources tied up in other work). SK .Net was released at v1.0 first in December of 2023, and then Python/Java released v1.0 at Microsoft Build in May 2024. Since Build 2024, Python has been rapidly releasing new features to get to parity with .Net. Are there specific features you see in SK .Net that aren't available in Python? 

Some notes:

- The agent framework is available in both Python and .Net, with minor feature divergence -- Python has work items to support ChatHistoryReducer as well as supporting Kernel Args/Prompt Templates for Agents.
- The process framework is in both Python and .Net, with minor feature divergence -- Python has in-progress work for state management and the map step.
- The new memory store abstractions/vector search/text search capabilities are available in both Python and .Net -- both languages still have some memory connectors that need to be moved to the new abstractions.
- AI service connectors are at parity in both Python and .Net. 
- Content types are at parity in both Python and .Net.
- Filters are at parity between Python and .Net.
- Telemetry is at parity (or close-to) between Python and .Net.

As mentioned, please do provide us the functionality that's missing in Python as compared to .Net so we can make sure we're tracking the items in our public backlog. Thank you!

**Reply by lovedeepatsgit** [reply]
I Have seen capabilities like 'Text Search Function Calling' https://learn.microsoft.com/en-us/semantic-kernel/concepts/text-search/text-search-function-calling?pivots=programming-language-python, Text Search Plugins which are available in .Net but not in python.

But Thanks a Lot @moonbox3 these insights are more than enough to choose SK for production.

**Reply by moonbox3** [reply]
We are working to improve the Learn site docs, and sometimes the code gets released and the learn site doc update isn't as fast of a follow as it should be. This is a great of example, and I will make a note that we need to get the Python pivot updated as the code is available. You can find the code sample here: https://github.com/microsoft/semantic-kernel/blob/main/python/samples/concepts/search/bing_text_search_as_plugin.py. There are some other search concept examples available, too.

**Reply by lovedeepatsgit** [reply]
Sure, @moonbox3  I will check this out. Are we getting something like Magentic-One, also in Semantic Kernels?

**Reply by moonbox3** [reply]
SK's Process Framework can be used to replicate operations and orchestration similar to Magentic-One. As we continue collaborating with the AutoGen team to plan its integration with SK, we will identify and address any gaps between the capabilities provided by Magentic-One and those currently supported by the SK Process Framework.
